### PR TITLE
Nina stiefel last

### PR DESCRIPTION
**Pull request by ninamiolane**
Lifting and retraction maps for stiefel manifolds

### CODE REVIEW

File: geomstats/stiefel.py [hunk 1]
```diff
@@ -66,3 +70,284 @@ def random_uniform(self, n_samples=1):
...
+                dimension=dimension,
+                signature=(dimension, 0, 0))
+        self.embedding_metric = EuclideanMetric(n*p)
+
+    def inner_product(self, tangent_vec_a, tangent_vec_b, base_point=None):
+        """
+        Compute the Frobenius inner product of tangent_vec_a and tangent_vec_b
+        at base_point using the metric of the embedding space.
+        """
+        # TODO(nina): This is for the euclidean metric, not canonical. change.
```
Comment (johmathe) [inline geomstats/stiefel.py]: so this is wrong then?

File: geomstats/stiefel.py [hunk 2]
```diff
@@ -66,3 +70,284 @@ def random_uniform(self, n_samples=1):
...
+
+        matrix_rd = gs.matmul(matrix_r, gs.transpose(matrix_d, axes=(0, 2, 1)))
+        matrix_v[:, :, p:2*p] = gs.matmul(matrix_v[:, :, p:2*p], matrix_rd)
+        matrix_v = gs.concatenate(
+            [gs.concatenate([matrix_m, matrix_n], axis=1),
+             matrix_v[:, :, p:2*p]],
+            axis=2)
+
+        for k in range(max_iter):
+
```
Comment (johmathe) [inline geomstats/stiefel.py]: line break - in general please make sure they are clean everywhere

File: geomstats/stiefel.py [hunk 3]
```diff
@@ -66,3 +70,284 @@ def random_uniform(self, n_samples=1):
...
+        # Q, R = gs.linalg.qr(base_point + tangent_vec)
+        # TODO: remove cycle, when qr will be vectorized
+        matrix_q = gs.zeros_like(base_point)
+        matrix_r = gs.zeros((n_retractions, p, p))
+
+        for i, k in enumerate(base_point + tangent_vec):
+            matrix_q[i], matrix_r[i] = gs.linalg.qr(k)
+
+        # flipping signs
+        # Q = gs.matmul(Q, gs.diag(gs.sign(gs.sign(gs.diagonal(R)) + 0.5)))
```
Comment (johmathe) [inline geomstats/stiefel.py]: why is this still commented? Either get rid of it or uncomment -

File: geomstats/stiefel.py [hunk 4]
```diff
@@ -66,3 +70,284 @@ def random_uniform(self, n_samples=1):
...
+        assert (n_tangent_vecs == n_base_points
+                or n_tangent_vecs == 1
+                or n_base_points == 1)
+
+        if n_base_points == 1:
+            base_point = gs.tile(base_point, (n_tangent_vecs, 1, 1))
+        if n_tangent_vecs == 1:
+            tangent_vec = gs.tile(tangent_vec, (n_base_points, 1, 1))
+        n_retractions = gs.maximum(n_base_points, n_tangent_vecs)
+        # Q, R = gs.linalg.qr(base_point + tangent_vec)
```
Comment (johmathe) [inline geomstats/stiefel.py]: why is this commented? Uncomment or add TODO

Comment (johmathe) [reply geomstats/stiefel.py]: or delete :)

Comment (ninamiolane) [reply geomstats/stiefel.py]: Done.

File: geomstats/stiefel.py (hunk 3 above)
Comment (ninamiolane) [reply geomstats/stiefel.py]: Done.

File: geomstats/stiefel.py (hunk 2 above)
Comment (ninamiolane) [reply geomstats/stiefel.py]: Done.

**Pull request merged**